packages/react/src/components/NumberInput/NumberInput.tsx

```diff
@@ -319,17 +431,30 @@ const NumberInput = React.forwardRef<HTMLInputElement, NumberInputProps>(
         return;
       }
 
-      const state = {
-        value:
-          allowEmpty && event.target.value === ''
-            ? ''
-            : Number(event.target.value),
-        direction: value < event.target.value ? 'up' : 'down',
-      };
-      setValue(state.value);
-
-      if (onChange) {
-        onChange(event, state);
+      if (type === 'number') {
+        const state = {
+          value:
+            allowEmpty && event.target.value === ''
+              ? ''
+              : Number(event.target.value),
+          direction: value < event.target.value ? 'up' : 'down',
+        };
+        setValue(state.value);
+
+        if (onChange) {
+          onChange(event, state);
+        }
+        return;
+      }
+
+      if (type === 'text') {
+        const _value =
+          allowEmpty && event.target.value === '' ? '' : event.target.value;
+        setNumberValue(numberParser.parse(_value));
+        setInputValue(_value);
+        // The onChange prop isn't called here because it will be called on blur
+        // or on click of a stepper, after the number is parsed and formatted
+        // according to the locale.
```

Comment:
Same thing here - I know this is verbose and a bit duplicative but trying to clearly separate these code paths to avoid confusion and potential regressions.